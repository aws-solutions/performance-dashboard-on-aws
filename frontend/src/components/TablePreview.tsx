import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

type Props = {
  title: string;
  headers: Array<string>;
  data?: Array<object>;
};

const getKeyValue = (key: string) => (obj: Record<string, any>) => obj[key];

const TablePreview = (props: Props) => {
  return (
    <div>
      <h2 className="margin-left-3">{props.title}</h2>
      <table className="usa-table usa-table--borderless margin-left-3">
        <thead>
          <tr>
            {props.headers.length
              ? props.headers.map((header, index) => {
                  return (
                    <th key={index} scope="col">
                      <div className="display-inline">{header}</div>
                      <div className="grid-row flex-column display-inline-block height-1 text-center margin-left-1">
                        <div className="grid-col flex-6 height-1">
                          <FontAwesomeIcon icon="caret-up" size="sm" />
                        </div>
                        <div className="grid-col flex-6">
                          <FontAwesomeIcon icon="caret-down" size="sm" />
                        </div>
                      </div>
                    </th>
                  );
                })
              : null}
          </tr>
        </thead>
        <tbody>
          {props.data &&
            props.data.length &&
            props.data.map((data, indexData) => {
              return (
                <tr key={indexData}>
                  {props.headers.map((header, index) => {
                    return index === 0 ? (
                      <th key={index} scope="row">
                        {getKeyValue(header)(data)}
                      </th>
                    ) : (
                      <td key={index}>{getKeyValue(header)(data)}</td>
                    );
                  })}
                </tr>
              );
            })}
        </tbody>
      </table>
    </div>
  );
};

export default TablePreview;
