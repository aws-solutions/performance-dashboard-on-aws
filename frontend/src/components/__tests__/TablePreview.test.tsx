import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import TablePreview from "../TablePreview";

import { library } from "@fortawesome/fontawesome-svg-core";
import { faCaretUp, faCaretDown } from "@fortawesome/free-solid-svg-icons";

library.add(faCaretUp, faCaretDown);

test("renders the title of the table preview component", async () => {
  const { getByText } = render(
    <TablePreview title="test title" headers={["test"]} />,
    { wrapper: MemoryRouter }
  );
  expect(getByText("test title")).toBeInTheDocument();
});
