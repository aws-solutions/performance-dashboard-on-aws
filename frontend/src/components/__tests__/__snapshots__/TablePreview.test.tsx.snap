// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`table preview should match snapshot 1`] = `
<div>
  <div>
    <h2
      class="margin-left-3"
    >
      test title
    </h2>
    <table
      class="usa-table usa-table--borderless margin-left-3"
    >
      <thead>
        <tr>
          <th
            scope="col"
          >
            <div
              class="display-inline"
            >
              test
            </div>
            <div
              class="grid-row flex-column display-inline-block height-1 text-center margin-left-1"
            >
              <div
                class="grid-col flex-6 height-1"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-caret-up fa-w-10 fa-sm "
                  data-icon="caret-up"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 320 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M288.662 352H31.338c-17.818 0-26.741-21.543-14.142-34.142l128.662-128.662c7.81-7.81 20.474-7.81 28.284 0l128.662 128.662c12.6 12.599 3.676 34.142-14.142 34.142z"
                    fill="currentColor"
                  />
                </svg>
              </div>
              <div
                class="grid-col flex-6"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-caret-down fa-w-10 fa-sm "
                  data-icon="caret-down"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 320 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                    fill="currentColor"
                  />
                </svg>
              </div>
            </div>
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</div>
`;
